crates/core_simd/src/vector.rs

```diff
@@ -176,34 +176,62 @@ where
         unsafe { &mut *(self as *mut Self as *mut [T; N]) }
     }
 
+    /// Load a vector from an array of `T`.
+    ///
+    /// This function is necessary since `repr(simd)` has padding for non-power-of-2 vectors (at the time of writing).
+    /// With padding, `read_unaligned` will read past the end of an array of N elements.
+    ///
+    /// # Safety
+    /// Reading `ptr` must be safe, as if by `<*const [T; N]>::read_unaligned`.
+    const unsafe fn load(ptr: *const [T; N]) -> Self {
+        let mut tmp = core::mem::MaybeUninit::uninit();
+        // SAFETY: `Simd<T, N>` always contains `N` elements of type `T`.  It may have padding
+        // which does not need to be initialized.  The safety of reading `ptr` is ensured by the
+        // caller.
+        unsafe {
+            core::ptr::copy_nonoverlapping(ptr, tmp.as_mut_ptr() as *mut _, 1);
+            tmp.assume_init()
+        }
+    }
+
+    /// Store a vector to an array of `T`.
+    ///
+    /// See `load` as to why this function is necessary.
+    ///
+    /// # Safety
+    /// Writing to `ptr` must be safe, as if by `<*mut [T; N]>::write_unaligned`.
+    const unsafe fn store(self, ptr: *mut [T; N]) {
```

Comment:
I think you'll want to explicitly copy `self` to a temporary first, and then memcpy from the temporary, since afaict llvm optimizes that to a vector load/store due to the vector insns generated by the temporary, whereas just calling memcpy generates only a memcpy so llvm doesn't see any vector operations and generates possibly less-efficient code.

compare `store` vs. `store2` -- `store2` has the temporary.
https://godbolt.org/z/jWsc6ezP4

Reply:
Isn't it *intentional*, though, that it's not a vector read?  Because I read the OP as saying that that would generate an over-long read.

Reply:
no, because LLVM IR defines vector `load`/`store` instructions to never read/write the padding, as i explained on Zulip. `read_unaligned` has no such guarantee

Reply:
so we *intentionally* want a llvm `load`/`store` instruction with `align <elem-align>` not full-simd-type-align

Reply:
It's not written as a vector store, because that would be wrong (writing past the end of the array), but it should lower as one.  I added the temporary with a comment as to why it's necessary.

Reply:
Oh, I see, I was mixing up the LLVM type and the Rust type.